src/services.py

```diff
@@ -143,3 +158,66 @@ def __init__(self, Security=Security):
     def get_all(self):
         with session_scope() as session:
             return [sec.asdict() for sec in session.query(self.Security).all()]
+
+
+class LinkedinService:
+    def __init__(self, User=User, UserService=UserService):
+        self.User = User
+        self.UserService = UserService
+
+    @validate_input(LINKEDIN_BUYER_PRIVILEGES_SCHEMA)
+    def activate_buyer_privileges(self, code, redirect_uri, user_email):
+        user_data = self.get_user_data(code=code, redirect_uri=redirect_uri)
+        is_email = self.is_email_matched(
+            linkedin_email=user_data.get("email"), user_email=user_email
+        )
+        if is_email:
+            user = self.UserService().get_user_by_email(email=user_email)
+            return self.UserService().activate_buy_privileges(user_id=user.get("id"))
+        else:
+            raise InvalidRequestException("Linkedin email does not match")
+
+    @validate_input(LINKEDIN_CODE_SCHEMA)
+    def get_user_data(self, code, redirect_uri):
+        token = self.get_token(code=code, redirect_uri=redirect_uri)
+        return self.get_user_email(**token)
+
+    @validate_input(LINKEDIN_CODE_SCHEMA)
+    def get_token(self, code, redirect_uri):
+        token = requests.post(
+            "https://www.linkedin.com/oauth/v2/accessToken",
+            headers={"Content-Type": "x-www-form-urlencoded"},
+            params={
+                "grant_type": "authorization_code",
+                "code": code,
+                "redirect_uri": redirect_uri,
+                "client_id": APP_CONFIG.get("CLIENT_ID"),
+                "client_secret": APP_CONFIG.get("CLIENT_SECRET"),
+            },
+        ).json()
+        return {"token": token.get("access_token")}
+
+    @validate_input(LINKEDIN_TOKEN_SCHEMA)
+    def get_user_profile(self, token):
+        user_profile = requests.get(
+            "https://api.linkedin.com/v2/me",
+            headers={"Authorization": f"Bearer {token}"},
+        ).json()
+        first_name = user_profile.get("localizedFirstName")
+        last_name = user_profile.get("localizedLastName")
+        return {"full_name": f"{first_name} {last_name}"}
+
+    @validate_input(LINKEDIN_TOKEN_SCHEMA)
+    def get_user_email(self, token):
+        email = requests.get(
+            "https://api.linkedin.com/v2/emailAddress?q=members&projection=(elements*(handle~))",
+            headers={"Authorization": f"Bearer {token}"},
+        ).json()
+        return {"email": email.get("elements")[0].get("handle~").get("emailAddress")}
```

Comment:
Why not just return the email directly, why need to wrap it in a dict? My main intention with wrapping some of the return values with dict is so that it can be directly serialised by the API as JSON. But if it's an internal function I don't see the point.